iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/IcebergAcidUtil.java

```diff
@@ -0,0 +1,135 @@
+  }
+
+  private static final Types.NestedField PARTITION_STRUCT_META_COL = null; // placeholder value in the map
+  private static final Map<Types.NestedField, Integer> DELETE_FILE_READ_META_COLS = Maps.newLinkedHashMap();
+
+  static {
+    DELETE_FILE_READ_META_COLS.put(MetadataColumns.SPEC_ID, 0);
+    DELETE_FILE_READ_META_COLS.put(PARTITION_STRUCT_META_COL, 1);
+    DELETE_FILE_READ_META_COLS.put(MetadataColumns.FILE_PATH, 2);
+    DELETE_FILE_READ_META_COLS.put(MetadataColumns.ROW_POSITION, 3);
+  }
+
+  private static final Types.NestedField PARTITION_HASH_META_COL = Types.NestedField.required(
+      MetadataColumns.PARTITION_COLUMN_ID, MetadataColumns.PARTITION_COLUMN_NAME, Types.LongType.get());
+  private static final Map<Types.NestedField, Integer> DELETE_SERDE_META_COLS = Maps.newLinkedHashMap();
+
+  static {
+    DELETE_SERDE_META_COLS.put(MetadataColumns.SPEC_ID, 0);
+    DELETE_SERDE_META_COLS.put(PARTITION_HASH_META_COL, 1);
+    DELETE_SERDE_META_COLS.put(MetadataColumns.FILE_PATH, 2);
+    DELETE_SERDE_META_COLS.put(MetadataColumns.ROW_POSITION, 3);
+  }
+
+  /**
+   * @param dataCols The columns of the original file read schema
+   * @param table The table object - it is used for populating the partition struct meta column
+   * @return The schema for reading files, extended with metadata columns needed for deletes
+   */
+  public static Schema createFileReadSchemaForDelete(List<Types.NestedField> dataCols, Table table) {
+    List<Types.NestedField> cols = Lists.newArrayListWithCapacity(dataCols.size() + DELETE_FILE_READ_META_COLS.size());
+    DELETE_FILE_READ_META_COLS.forEach((metaCol, index) -> {
+      if (metaCol == PARTITION_STRUCT_META_COL) {
+        cols.add(MetadataColumns.metadataColumn(table, MetadataColumns.PARTITION_COLUMN_NAME));
+      } else {
+        cols.add(metaCol);
+      }
+    });
+    cols.addAll(dataCols);
+    return new Schema(cols);
+  }
+
+  /**
+   * @param dataCols The columns of the serde projection schema
+   * @return The schema for SerDe operations, extended with metadata columns needed for deletes
+   */
+  public static Schema createSerdeSchemaForDelete(List<Types.NestedField> dataCols) {
+    List<Types.NestedField> cols = Lists.newArrayListWithCapacity(dataCols.size() + DELETE_SERDE_META_COLS.size());
+    DELETE_SERDE_META_COLS.forEach((metaCol, index) -> cols.add(metaCol));
+    cols.addAll(dataCols);
+    return new Schema(cols);
+  }
+
+  public static PositionDelete<Record> getPositionDelete(Schema schema, Record rec) {
+    PositionDelete<Record> positionDelete = PositionDelete.create();
+    String filePath = rec.get(DELETE_SERDE_META_COLS.get(MetadataColumns.FILE_PATH), String.class);
+    long filePosition = rec.get(DELETE_SERDE_META_COLS.get(MetadataColumns.ROW_POSITION), Long.class);
+
+    int dataOffset = DELETE_SERDE_META_COLS.size(); // position in the rec where the actual row data begins
+    Record rowData = GenericRecord.create(schema);
```

Comment:
Creating this is costly for every row - we might want to keep a record object, and reuse it.

Reply:
Shall we keep a cache with a Record object for each schema?

Reply:
We could use a caffeine cache, but I'm not sure how much cost is associated with a potential cache miss

Reply:
Added a cache, let me know what you think